keel-core/src/main/kotlin/com/netflix/spinnaker/keel/serialization/SubmittedEnvironmentDeserializer.kt

```diff
@@ -0,0 +1,62 @@
+package com.netflix.spinnaker.keel.serialization
+
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.InjectableValues
+import com.fasterxml.jackson.databind.JsonMappingException
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.deser.std.StdNodeBasedDeserializer
+import com.fasterxml.jackson.module.kotlin.convertValue
+import com.netflix.spinnaker.keel.api.Constraint
+import com.netflix.spinnaker.keel.api.NotificationConfig
+import com.netflix.spinnaker.keel.api.SubmittedEnvironment
+import com.netflix.spinnaker.keel.api.SubmittedResource
+import com.netflix.spinnaker.keel.api.SubnetAwareLocations
+import java.lang.IllegalArgumentException
+
+/**
+ * Deserializer that allows us to propagate values such as [SubmittedEnvironment.locations] to all
+ * resources in the environment without having to make the corresponding properties in the resource
+ * specs nullable and continually have to look up the environment.
+ */
+class SubmittedEnvironmentDeserializer : StdNodeBasedDeserializer<SubmittedEnvironment>(SubmittedEnvironment::class.java) {
+  override fun convert(root: JsonNode, context: DeserializationContext): SubmittedEnvironment {
+    with(context) {
+      val name = root.path("name").textValue()
+      val constraints = try {
+        mapper
+          .convertValue<Set<Constraint>?>(root.path("constraints"))
+          ?: emptySet()
+      } catch (e: IllegalArgumentException) {
+        throw JsonMappingException.wrapWithPath(e, root, "constraints")
+      }
+      val notifications = try {
+        mapper
+          .convertValue<Set<NotificationConfig>?>(root.path("notifications"))
+          ?: emptySet()
+      } catch (e: IllegalArgumentException) {
+        throw JsonMappingException.wrapWithPath(e, root, "notifications")
+      }
+      val locations = try {
+        mapper
+          .convertValue<SubnetAwareLocations?>(root.path("locations"))
+      } catch (e: IllegalArgumentException) {
+        throw JsonMappingException.wrapWithPath(e, root, "locations")
+      }
+      val resources = try {
+        mapper.copy().let { nestedMapper ->
+          nestedMapper.injectableValues = InjectableValues.Std(mapOf("locations" to locations))
+          nestedMapper
+            .convertValue<Set<SubmittedResource<*>>?>(root.path("resources"))
+            ?: emptySet()
+        }
+      } catch (e: IllegalArgumentException) {
+        throw JsonMappingException.wrapWithPath(e, root, "resources")
+      }
```

Comment:
Here's the interesting bit. I read the `locations` property (if there is one) on the `SubmittedEnvironment` then create a nested `ObjectMapper` with that injected value and read the set of `SubmittedResource`s. The `locations` are only in scope while I'm doing that and don't leak out into the wider deserialization context.

Reply:
I've tidied the code up a bit now so there are no so many try/catch blocks everywhere. Intent remains identical.